src/QsCompiler/CompilationManager/DataStructures.cs

```diff
@@ -41,69 +50,357 @@ internal class CodeLine
 
         internal readonly int Indentation; // Note: This denotes the initial indentation at the beginning of the line
         internal readonly ImmutableArray<int> ExcessBracketPositions;
+        internal readonly ImmutableArray<int> ErrorDelimiterPositions;
 
         // convention: first one opens a string, and we have the usual inclusions: {0,5} means the chars 0..4 are the content of a string
         // -> i.e. -1 means the string starts on a previous line, and an end delimiter that is equal to Text.Length means that the string continues on the next line
-        internal readonly ImmutableArray<int> StringDelimiters; // note that this property is only properly initialized *after* knowing the sourrounding lines
+        internal readonly ImmutableArray<int> StringDelimiters; // note that this property is only properly initialized *after* knowing the surrounding lines
+
+        internal readonly StringContext BeginningStringContext;
+        internal readonly StringContext EndingStringContext;
+
+        public CodeLine(
+            string text,
+            string lineEnding,
+            string withoutEnding,
+            string? endOfLineComment,
+            int indentation,
+            IEnumerable<int> excessBracketPositions,
+            IEnumerable<int> errorDelimiterPositions,
+            IEnumerable<int> stringDelimiters,
+            StringContext beginningStringContext,
+            StringContext endingStringContext)
+        {
+            this.Text = text;
+            this.LineEnding = lineEnding;
+            this.WithoutEnding = withoutEnding;
+            this.EndOfLineComment = endOfLineComment;
+            this.Indentation = indentation;
+            this.ExcessBracketPositions = excessBracketPositions.ToImmutableArray();
+            this.ErrorDelimiterPositions = errorDelimiterPositions.ToImmutableArray();
+            this.StringDelimiters = stringDelimiters.ToImmutableArray();
+            this.BeginningStringContext = beginningStringContext;
+            this.EndingStringContext = endingStringContext;
+
+            ScopeTracking.VerifyStringDelimiters(text, stringDelimiters);
+            ScopeTracking.VerifyExcessBracketPositions(this, excessBracketPositions);
+        }
+
+        private CodeLine(StringContext beginningStringContext)
```

Comment:
As there're multiple constructors with different arguments it would be nice to have comments on how they are different.

Reply:
That's a good idea. I realize I made a lot more constructors in this PR so adding comments to help sort out which ones are for what is good. I've added descriptions for these constructors to the code to help make things a bit cleaner to understand.